include/aws/common/byte_order.h

```diff
@@ -0,0 +1,102 @@
+#ifndef AWS_COMMON_BYTE_ORDER_H
+#define AWS_COMMON_BYTE_ORDER_H
+
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <stdint.h>
+
+/**
+ * Returns 1 if machine is big endian, 0 if little endian.
+ * If you compile with even -O1 optimization, this check is completely optimized out at compile
+ * time and code which calls "if (is_big_endian())" will do the right thing without branching.
+ */
+static inline int is_big_endian() {
+    const uint16_t z = 0x100;
+    return *(const uint8_t*)&z;
+}
+
+/**
+ * Swaps bytes of 64 bit integer.
+ */
+static inline uint64_t byteswap64(uint64_t x) { return
+        (x << 56) | (x >> 56) |
+        ((x & 0x0000ff00ULL) << 40) | ((x >> 40) & 0x0000ff00ULL) |
+        ((x & 0x00ff0000ULL) << 24) | ((x >> 24) & 0x00ff0000ULL) |
+        ((x & 0xff000000ULL) <<  8) | ((x >>  8) & 0xff000000ULL);
+}
+
+/**
+ * Swaps bytes of 32 bit integer.
+ */
+static inline uint32_t byteswap32(uint32_t x) { return
```

Comment:
This will be significantly slower than `htonl`/`ntohl` on x86. x86 has specific assembly instructions for byteswapping 16-bit (`ROR $8, %ax`), 32-bit (`BSWAP %eax`) and 64-bit (`BSWAP %rax`) registers.